### DISCUSSION TITLE
Using app.dependency_overrides with parametrized depenencies

### DISCUSSION THREAD

**Discussion by shohamy7** [fastapi/fastapi#6834] [Questions] [answered]
Hi! 
I'm trying to test my FastAPI application and got into some trouble with `app.dependency_overrides`.
I'm using parametrized dependencies and when I'm trying to add the dependency to the dependency_overrides dict and run my tests, the tests runs with the original dependency instead of the mock dependency I created.
When I use regular function, it works fine. The problem occurs only with class dependencies.
Here what I tried to do: 
```
class NameDependency:
    def __init__(self, name: str):
        self._name = name
    def __call__(self) -> str:
        return self._name

app = FastAPI()

@app.get('/')
async def hello(name: str = Depends(NameDependency('foo'))):
    return {'message': f'hello {name}'}

client = TestClient(app)
app.dependency_overrides[NameDependency] = Mock(return_value='bar')

def test_hello():
    res = client.get('/')
    assert res.json()['message'] == 'hello bar'
```
When I run this test, instead of passing, I get AssertionError hello foo == hello bar.
I also tried the following alternatives:
```
# 1st alternative
app.dependency_overrides[NameDependency('foo')] = Mock(return_value='bar')

# 2nd alternative
app.dependency_overrides[NameDependency] = Mock(return_value=Mock(return_value='bar'))
```
I've tried more but I could not find the correct way to do that. 
I think I'm missing something and I'll glad if someone can help me. I didn't find an example to this online and I didn't see this in the docs either. 

Thanks for anyone that will help!

**Comment by Mause** [comment, marked as answer]
It's because your NameDependency class isn't hashable (so the lookups in the dependency_overrides dict is failing).
It's easy to fix with dataclasses:

```py
from fastapi import FastAPI, Depends
from fastapi.testclient import TestClient
from dataclasses import dataclass


@dataclass(frozen=True, eq=True)
class NameDependency:
    name: str

    def __call__(self) -> str:
        return self.name


app = FastAPI()


@app.get("/")
async def hello(name: str = Depends(NameDependency("foo"))):
    return {"message": f"hello {name}"}


client = TestClient(app)
app.dependency_overrides[NameDependency("foo")] = lambda: "bar"


def test_hello():
    res = client.get("/")
    assert res.json()["message"] == "hello bar"
```

**Comment by shohamy7** [comment]
That's works, Thank you!
I wish this would have been written in the docs, it would save me a lot of time :)

**Comment by delijati** [comment]
> It's because your NameDependency class isn't hashable (so the lookups in the dependency_overrides dict is failing).
> It's easy to fix with dataclasses:
> 
> ```python
> from fastapi import FastAPI, Depends
> from fastapi.testclient import TestClient
> from dataclasses import dataclass
> 
> 
> @dataclass(frozen=True, eq=True)
> class NameDependency:
>     name: str
> 
>     def __call__(self) -> str:
>         return self.name
> 
> 
> app = FastAPI()
> 
> 
> @app.get("/")
> async def hello(name: str = Depends(NameDependency("foo"))):
>     return {"message": f"hello {name}"}
> 
> 
> client = TestClient(app)
> app.dependency_overrides[NameDependency("foo")] = lambda: "bar"
> 
> 
> def test_hello():
>     res = client.get("/")
>     assert res.json()["message"] == "hello bar"
> ```

Took me some time to find this ... Should definitely go to the docs or faq. Thanks a lot @Mause 
To get this working on any `class` you have to overwrite `__eg__` and `__hash__`:

```python
class JWTBearer(HTTPBearer):
    def __init__(self, auto_error: bool = True):
        super().__init__(auto_error=auto_error)

    def __hash__(self):
        # FIXME find something uniq and repeatable                                     
        return 1234   
   
    def __eq__(self, other):                       
        """Overrides the default implementation"""
        if isinstance(other, JWTBearer):
            return self.auto_error == other.auto_error                  
        return False
```

**Comment by alijewers** [comment]
Would be really great if this could make it into the documentation, our team has ended up back here several times now after encountering this problem independently

**Comment by rcbop** [comment]
How can I override the dependency when I have the following mutable dependency?
I tried making it a frozen dataclasse but I get an error because of the roles parameter.

I also tried implementing __hash__ and __eq__. Will I need one override for each combination of roles?

```
@router.get("/", response_model.status.HTTP_200_OK)
def my_endpoint(current_user: User = Depends(CheckRoles(roles=['admin', 'visitor']))):
```

```
class CheckRoles:
    def __init__(self, roles) -> None:
        self.roles = roles

    def __call__(self,
                 settings: Annotated[Settings, Depends(get_settings)],
                 authorization_jwt: Optional[str] = Header(None),
                 user_repository: UserRepository = Depends()) -> bool:
        ...
```
